<a href="https://colab.research.google.com/github/OlegV12/Stats/blob/Lesson_7/mathStats.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from math import factorial, pow, e, pi

In [5]:
import numpy as np
from scipy.stats import norm
import pandas as pd

In [ ]:

def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [ ]:
def bernulli(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  return combinations(n, k) * pow(p, k) * pow((1-p), (n-k))

In [ ]:
def poisson(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  a = n * p # lambda

  return pow(a, k) / factorial(k) * pow(e, (-a))



In [ ]:
def mean_sqrt_dev(a):
  res = []
  mean = sum(a) / len(a)
  for i in a:
    res.append(pow((i - mean), 2))
  return np.sqrt(sum(res) / len(a)), np.sqrt(sum(res) / (len(a)-1))


In [ ]:
def prob(mu, x, sqd):
  return (x-mu)/sqd

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [132]:
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])


In [133]:
X = X.reshape(len(X), 1)
y = y.reshape(len(y), 1)

In [134]:
X = np.hstack([np.ones((10, 1)), X])
X

array([[  1.,  35.],
       [  1.,  45.],
       [  1., 190.],
       [  1., 200.],
       [  1.,  40.],
       [  1.,  70.],
       [  1.,  54.],
       [  1., 150.],
       [  1., 120.],
       [  1., 110.]])

In [135]:
b = np.dot(np.linalg.inv(np.dot(X.T, X)), X.T@y)
b

array([[444.17735732],
       [  2.62053888]])

y = 444.177 + 2.62*x

In [30]:
# вариант без intercept
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
X = X.reshape(len(X), 1)
y = y.reshape(len(y), 1)
b = np.dot(np.linalg.inv(np.dot(X.T, X)), X.T@y)
b

array([[5.88982042]])

y = 5.8898*x

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [33]:
def mse (X, y, b1):
  return np.sum((b1*X-y)**2)/len(X)

In [ ]:
x = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
B1 = 1
alpha = 1e-6
for i in range(500):
  B1 -= alpha*(2/10)*np.sum((B1*x-y)*x) 
  print (B1, mse(x, y, B1))

Ответ: коэфицент 5.8898

В каких случаях для вычисления доверительных интервалов и проверки статистических гипотез используется таблица значений функции Лапласа, а в каких - таблица критических точек распределения Стьюдента? 

Для небольших выборок используется таблица критических точек распределения Стьюдента, при больших колличествах наблюдений показания в таблицах равны. 

Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

In [150]:
x = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
a = 20
b = 20

alpha = 1e-5
for i in range(2000000):

  a_temp = a
  b_temp = b

  a -= alpha*(2/10)*np.sum(((a_temp*x+b_temp)-y)*x) 

  b -= alpha*(2/10)*np.sum((a_temp*x + b_temp)-y)
 
print (a,b)

2.6206613935254026 444.1607127912527
